src/main/scala/firrtl/passes/AnnotateValidMemConfigs.scala

```diff
@@ -0,0 +1,298 @@
+  min: Int,
+  // step size
+  inc: Int,
+  max: Int,
+  // these values should not be used, regardless of min,inc,max
+  illegal: Option[List[Int]]
+){
+  def getValid = {
+    val range = (min to max by inc).toList
+    range.filterNot(illegal.getOrElse(List[Int]()).toSet)
+  }
+}
+
+case class MemDimension(
+  rules: Option[DimensionRules],
+  set: Option[List[Int]]
+){
+  require (
+    if(rules == None) set != None else set == None, 
+    "Should specify either rules or a list of valid options, but not both"
+  )
+  def getValid = set.getOrElse(rules.get.getValid).sorted
+}
+
+case class SRAMConfig(
+  ymux: String = "",
+  ybank: String = "",
+  width: Int,
+  depth: Int,
+  xsplit: Int = 1,
+  ysplit: Int = 1
+){
+  // how many duplicate copies of this SRAM are needed
+  def num = xsplit * ysplit
+  def serialize(pattern: String): String = {
+    val fieldMap = getClass.getDeclaredFields.map{f => 
+      f.setAccessible(true)
+      f.getName -> f.get(this)
+    }.toMap
+
+    val fieldDelimiter = """\[.*?\]""".r
+    val configOptions = fieldDelimiter.findAllIn(pattern).toList
+
+    configOptions.foldLeft(pattern)((b,a) => {
+      // Expects the contents of [] are valid configuration fields (otherwise key match error)
+      val fieldVal = {
+        try fieldMap(a.substring(1,a.length-1)) 
+        catch { case e: Exception => error("**SRAM config field incorrect**") }
+      }
+      b.replace(a,fieldVal.toString)
+    })
+  }
+}
+
+// Ex: https://www.ece.cmu.edu/~ece548/hw/hw5/meml80.pdf
+case class SRAMRules(
+  // column mux parameter (for adjusting aspect ratio)
+  ymux: (Int,String),
+  // vertical segmentation (banking -- tradeoff performance / area)
+  ybank: (Int,String),
+  width: MemDimension,
+  depth: MemDimension
+){
+  def getValidWidths = width.getValid
+  def getValidDepths = depth.getValid
+  def getValidConfig(width: Int, depth: Int): Option[SRAMConfig] = {
+    if (getValidWidths.contains(width) && getValidDepths.contains(depth)) 
+      Some(SRAMConfig(ymux = ymux._2, ybank = ybank._2, width = width, depth = depth))
+    else
+      None
+  }
+  def getValidConfig(m: DefMemory): Option[SRAMConfig] = getValidConfig(bitWidth(m.dataType).intValue,m.depth) 
+} 
+
+case class WMaskArg(
+  t: String,
+  f: String
+)
+
+// vendor-specific compilers
+case class SRAMCompiler(
+    vendor: String,
+    node: String,
+    // i.e. RF, SRAM, etc.
+    memType: String,
+    portType: String,
+    wMaskArg: Option[WMaskArg],
+    // rules for valid SRAM flavors
+    rules: Seq[SRAMRules],
+    // path to executable 
+    path: Option[String],
+    // (output) config file path
+    configFile: Option[String],
+    // config pattern
+    configPattern: Option[String],
+    // read documentation for details 
+    defaultArgs: Option[String],
+    // default behavior (if not used) is to have wmask port width = datawidth/maskgran
+    // if true: wmask port width pre-filled to datawidth
+    fillWMask: Boolean
+){
```

Comment:
This should go on the previous line.
ie. 
```scala
    fillWMask: Boolean){
```
See http://docs.scala-lang.org/style/declarations.html#classes